### ISSUE TITLE
Mongoose schema setter not being called when using $inc

### ISSUE THREAD

**Issue by nadirzebiri** [Automattic/mongoose#13158]
### Prerequisites

- [X] I have written a descriptive issue title

### Mongoose version

^6.9.0

### Node.js version

18.12.1

### MongoDB version

5.0.15

### Operating system

Windows

### Operating system version (i.e. 20.04, 11.3, 10)

11

### Issue

I run in a problem. I have defined a setter in my model:

```js
const userSchema = new mongoose.Schema({
    money: {
        balance: {
            type: Number,
            set: (v) => Number(v)?.toFixed(4),
            default: 0,
        },
    },
});
```

When I try increasing balance of a user, it does format the number when the value is positive but sometimes not when its negative, which is weird.

```js
await User.findByIdAndUpdate(userId, {
        $inc: {
            'money.balance': earnings,
        },
    },
    { runValidators: true }
);
```

Thanks for helping me

**Comment by IslandRhythms** [comment]
```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
  money: {
      balance: {
          type: Number,
          set: (v) => Number(v)?.toFixed(4),
          default: 0,
      },
  },
});

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    money: { balance: -4}
  });
  const doc = await Test.findOne();
  console.log(doc.money.balance);
  const update = await Test.findOneAndUpdate({}, {$inc: { 'money.balance': 1.5432 }}, { runValidators: true, returnDocument: 'after' });
  console.log('what is update', update);
}

run();
```

output:
```
-4
what is update {
  money: { balance: -2.4568000000000003 },
  _id: new ObjectId("640f9a776f1f8ebed5209f59"),
  __v: 0
}

```

**Comment by vkarpov15** [comment]
@IslandRhythms the issue with your script is that you're rounding to 4 decimal places, and `1.5432` is already at 4 decimal places. And remember that adding/subtracting floating points can have some precision issues in MongoDB and in Node.

Round to 2 and your script works fine:

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
  money: {
      balance: {
          type: Number,
          set: (v) => Number(v)?.toFixed(2),
          default: 0,
      },
  },
});

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    money: { balance: -4}
  });
  const doc = await Test.findOne();
  console.log(doc.money.balance);
  const update = await Test.findOneAndUpdate({}, {$inc: { 'money.balance': 1.5432 }}, { runValidators: true, returnDocument: 'after' });
  console.log('what is update', update);
}

run();
```

Output:

```
-4
what is update {
  money: { balance: -2.46 },
  _id: new ObjectId("6411e8439161dff75582c9d4"),
  __v: 0
}
```

Also works fine with negative numbers:

```javascript
  const update = await Test.findOneAndUpdate({}, {$inc: { 'money.balance': -1.5432 }}, { runValidators: true, returnDocument: 'after' });
```

Output:

```
-4
what is update {
  money: { balance: -5.54 },
  _id: new ObjectId("6411e8fbadd0e470b3c77254"),
  __v: 0
}

```

However, I think the issue is that using the document `$inc()` method does **not** currently call setters, for example:

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
  money: {
      balance: {
          type: Number,
          set: (v) => Number(v)?.toFixed(2),
          default: 0,
      },
  },
});

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    money: { balance: -4}
  });
  const doc = await Test.findOne();
  console.log(doc.money.balance);
  doc.$inc('money.balance', -1.5432);
  await doc.save();
  const update = await Test.findOne({ _id: doc._id });
  console.log('what is update', update);
}

run();
```

Output:

```
what is update {
  money: { balance: -5.5432 },
  _id: new ObjectId("6411e9691dd9bd127222770b"),
  __v: 0
}
```

**Comment by nadirzebiri** [comment]
@vkarpov15, if the document $inc() method doesn't call setters, why does it call it in your first example?

Here's what the $inc returns in different cases:

-4 + 1.5432 = (1 decimal) -2.5 ✔️
-4 + 1.5432 = (2 decimals) -2.46 ✔️
-4 + 1.5432 = (3 decimals) -2.457 ✔️
-4 + 1.5432 = (4 decimals) -2.4568000000000003 ❌
-4 + 1.5432 = (5 decimals) -2.4568000000000003 ❌
-4 + 1.5432 = (6 decimals) -2.4568000000000003 ❌
-4 +- 1.5432 = (1 decimals) -5.5 ✔️
-4 +- 1.5432 = (2 decimals) -5.54 ✔️
-4 +- 1.5432 = (4 decimals) -5.5432 ✔️
-4 +-1.54322533 = (5 decimals) = -5.54323 ✔️

**Comment by vkarpov15** [comment]
@nadirzebiri `findOneAndUpdate()` with `$inc` is a different code path than `doc.$inc()`.

The reason why you're seeing `2.4568000000000003` is [binary floating point precision issues](https://thecodebarbarian.com/a-nodejs-perspective-on-mongodb-34-decimal.html)

**Comment by nadirzebiri** [comment]
@vkarpov15 ok so how not to have this floating error? $inc looks like calling setters but not in some cases. Does the error come from mongoose or MangoDB?

**Comment by vkarpov15** [comment]
@nadirzebiri we have a fix in #13178, will ship the fix later today.
